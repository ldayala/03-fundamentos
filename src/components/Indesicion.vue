<template>
  <img v-if="img" :src="img" alt="bg" />
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <input type="text" name="" id="" placeholder="Hazme una pregunta" v-model="question" />
    <p>Recuerda terminar la preguunta con un signo de (?).</p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name: "Indesicion",

  data() {
    return {
      question: null,
      answer: null,
      img: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Pensando...."; //mientras hago la peticion answer mostrara pensando...
        const { answer, image } = await fetch("https://yesno.wtf/api").then(
          (r) => r.json()
        );
        this.img = image;
        this.answer = answer;
        switch (this.answer) {
          case "yes":
            this.answer = "Si!!!";
            break;
          case "no":
            this.answer = "No!!!";
          // break;
          default:
            this.answer = "Quizas";
            break;
        }
      } catch (error) {
        console.log('Indesicion Component Error:',error);
        this.answer = "Error en la llamada a la Api";
        this.img = null;
      }
    },
  },
  watch: {
    //este es un observable, va estar pendiente de cualquier cambio en la propiedad que le pasemos
    question(value, oldvalue) {
      console.log("valor:", value);
      this.isValidQuestion = false;
      //valor actual , ultimo valor antes del cambio
      if (!value.includes("?")) return;

      this.isValidQuestion = true;
      //TODO: hacer peticion http
      this.getAnswer();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}

input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
