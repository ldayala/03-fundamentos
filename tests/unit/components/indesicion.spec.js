import { shallowMount } from "@vue/test-utils";
import Indesicion from "@/components/Indesicion";

describe("Indesicion Component", () => {
  let wrapper;
  let clgSpy;
  //hacemos un mock del fetch api para que lo pueda reconocer node ya que en este no existe
  global.fetch =
    jest.fn(() =>
      Promise.resolve({
        json: () =>
          Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif",
          }),
      })
    ) 

    beforeEach(() => {
      wrapper = shallowMount(Indesicion);
      //creo un espia para que me este espiando dentro del obj console el metodo log
      clgSpy = jest.spyOn(console, "log");
      jest.clearAllMocks(); // limpiamos todos los mocks
    });

  test("hacer match con el snatshop", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("escribir en el inpunt no debe disparar nada console.log()", async () => {
    //spy con la instancia de vue  wrapper.vm para poder espiar el getAnswer
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");
    await input.setValue("hola mundo");

    expect(clgSpy).toHaveBeenCalledTimes(1);
    //comprobamos que no se dispare el metodo
    //expect(getAnswerSpy).toHaveBeenCalledTimes(0)
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test('escribir "?" debe disparar el fetch', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");
    await input.setValue("hola?");
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  test("pruebas en el getAnswer", async() => {
    // ejecutamos el metodo getAnswer del componente
    await wrapper.vm.getAnswer()

    const img= wrapper.find('img');//cargamos el elemento img

    expect(img.exists()).toBeTruthy()//verificamos que exista el img
    // accedemos a la propiedad reactiva img y verficamos que tenga el valor pasado
    expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')

    expect(wrapper.vm.answer).toBe('Si!!!')

  });

  test("pruebas en getAnswer - Fallo de Api", async() => {

//TODO: simulacion fallo API
   fetch.mockImplementationOnce(()=>Promise.reject('Fallo Api'))

   await wrapper.vm.getAnswer()

    const img= wrapper.find('img')

    //verificamos que la imagen no exista
     expect(img.exists()).toBeFalsy()

     expect(wrapper.vm.answer).toBe('Error en la llamada a la Api')
  });
});
