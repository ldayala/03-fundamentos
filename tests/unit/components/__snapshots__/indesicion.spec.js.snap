// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indesicion Component hacer match con el snatshop 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" name="" id="" placeholder="Hazme una pregunta">
  <p>Recuerda terminar la preguunta con un signo de (?).</p>
  <!--v-if-->
</div>
`;
